proposals/0000-local-import.rst

```diff
@@ -0,0 +1,225 @@
+This is consistent with today's semantics for ``import``, which does not allow for:
+::
+
+  module MyModule where
+
+  import qualified Data.Set as Set
+  import Set
+
+As is already the case, module qualifiers can be re-used. For example, in:
+::
+
+  import Foo as A
+
+  main :: IO ()
+  main = do
+    import Bar as A
+    A.x
+
+The symbol ``x`` in ``A.x`` is searched in both ``Bar`` and ``Foo``.
+
+Finally, the syntactic shortcut ``Qualifier.{ <expression> }`` is introduced, which simply desugars to:
+::
+
+  let import Qualifier in <expression>
+The OCaml syntax, ``Qualifier.( … )``, was not chosen as it would steal syntax (this is currently parsed as a value constructor composed with the enclosed expression).
+   
+The following changes in the Haskell 2010 grammar are required:
+
+::
+
+  decls → { impdecls ; decl_1 ; … ; decl_n }    (n ≥ 0)
+        | { decl_1 ; … ; decl_n }
+  lexp  → …
+        | modid.{ lexp }
+        | do { stmts }
+        | do { impdecls ; stmts }
+
+
+Effect and Interactions
+-----------------------
+This proposal strictly extends the language, without affecting the behavior of existing code. It does not interact with any existing language extension.
+
+The changes give programmers various ways to reduce the number of toplevel imports, to limit their effect to specifics parts of the code and to convey intent about their uses. The shortcut syntax can be especially useful for scoping module imports over expressions with operators in DSLs.
+
+Furthermore, it encourages the use of qualified imports, as those can be locally “de-qualified” in order to improve readability.
+
+Costs and Drawbacks
+-------------------
+These changes should be easy to grasp by beginners. Moreover, they could greatly improve learnability of Haskell libraries, by expliciting the module each function comes from in examples and tutorials.
+
+Some existing tooling (e.g., ``snack``) assume that imports are only found at the toplevel and might be broken by this change. Perhaps more importantly, this change would make it harder for IDE-like tools such as ``hie`` to determine the set of valid completions ; such tools would need to be made context-sensitive, like OCaml's merlin.
+
+This change would make it harder to determine at first glance inter-dependencies between modules.
+
+Finally, some library writers might choose to design their library around this extension. Using such libraries without this extension enabled might be inconvenient, which could be perceived as a drawback by some users.
+
+Alternatives
+------------
+To our knowledge, there is no other language feature or extension providing similar benefits.
+
+It might be valuable to also allow some form of typelevel local import ; the shortcut syntax, in particular, could be used to simplify type signatures.
+
+The syntactic shortcut syntax is orthogonal to the rest of the proposal and could be entirely removed. The same functionality could potentially be achieved with QuasiQuoters.
```

Comment:
I have _some_ idea what this means, but it isn't clear.

- I'm guessing « the syntactic shortcut syntax » means « the `Module.{…}` syntax ». Say it like this (or something equivalent, anyway: no vague pointers).
- But do say that in Ocaml, it has prove very useful. Give examples of why.
- I don't understand what you mean about quasiquoters. But anyway: what would it look like? what would be the drawbacks?